### ISSUE TITLE
Master build is K.O. :  runtime error: left shift of 184 by 24 places cannot be represented in type 'int'

### ISSUE THREAD

**Issue by FdaSilvaYY** [openssl/openssl#11853]
See  https://travis-ci.org/github/openssl/openssl/builds/687826428

Config: 
```
EXTENDED_TEST="yes" CONFIG_OPTS="no-asm enable-ubsan enable-rc5 enable-md2 enable-ssl3 enable-ssl3-method enable-nextprotoneg no-shared enable-buildtest-c++ -fno-sanitize=alignment -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -Wno-unused-command-line-argument" CXX="clang++" 
```

Ouput :
```
read DSA key
crypto/pem/pvkfmt.c:42:18: runtime error: left shift of 184 by 24 places cannot be represented in type 'int'
../../util/wrap.pl ../../apps/openssl dsa -pubin -pubout -in msb-f.d -inform d -out msb-ff.dd -outform d => 1
    not ok 5 - d -> d

crypto/pem/pvkfmt.c:42:18: runtime error: left shift of 184 by 24 places cannot be represented in type 'int'
../../util/wrap.pl ../../apps/openssl dsa -pubin -pubout -in msb-f.d -inform d -out msb-ff.dp -outform p => 1
...
    not ok 8 - d -> p
    #   Failed test 'd -> p'
    #   at ../../test/recipes/tconversion.pl line 70.
```

**Comment by levitte** [comment]
Hmmm...  here's the code section where that happens:

https://github.com/openssl/openssl/blob/4fcd15c18ad6b5523a389863d3e5628d44db6eb4/crypto/pem/pvkfmt.c#L35-L45

Not a signed int in sight as far as I can see...  if you know how to fix it, PR welcome!

**Comment by t8m** [comment]
I think always casting `(unsigned int)*p++` should fix it. Well, it should be sufficient in the last case, but just for uniformity. Because the *p++ is promoted to int and then you try to shift it too much. The automatic conversion to unsigned int comes only after that.

**Comment by opensignature** [comment]
const unsigned char *p;
unsigned char hdr_buf[16];
hdr_buf[3] = 128;
p = hdr_buf;
a= read_ledword(&p);

runtime error: left shift of 128 by 24 places cannot be represented in type 'int'

**Comment by levitte** [comment]
I would to read the reasoning behind coercing the dereference of a pointer to an unsigned value to something signed. That makes zero sense to me...

**Comment by t8m** [comment]
That's C :grin: No need to seek reasoning. If you automatically convert something smaller (even if unsigned) to int and the "something smaller" fits, it is promoted to int and not unsigned int. And 184 fits into int.

**Comment by levitte** [comment]
Let's just say that C seems to become weirder as time goes :confused:

**Comment by kaduk** [comment]
> That's C  No need to seek reasoning. If you automatically convert something smaller (even if unsigned) to int and the "something smaller" fits, it is promoted to int and not unsigned int. And 184 fits into int.

To give chapter and verse, for C99 (well, n1256.pdf, the last free draft version), see section 6.3.1.1 clause 2

**Comment by bernd-edlinger** [comment]
I wonder why both competing PRs seem to fix other places where no runtime error happens?

**Comment by levitte** [comment]
Thank you @kaduk

**Comment by romen** [comment]
We should have probably flagged #11857 as `urgent`, as many PRs are failing on this!

**Comment by bernd-edlinger** [comment]
If it was "really" urgent the author should have split the urgent part out in a separate PR.
